io/subsystem/src/main/java/org/wildfly/extension/io/BufferPoolResourceDefinition.java

```diff
@@ -128,29 +119,46 @@ public void registerCapabilities(ManagementResourceRegistration resourceRegistra
     }
 
-    private static class BufferPoolAdd extends AbstractAddStepHandler {
+    @Override
+    public void registerAttributes(ManagementResourceRegistration resourceRegistration) {
+        resourceRegistration.registerReadOnlyAttribute(BUFFER_SIZE, new BufferReadAttributeHandler(BUFFER_SIZE));
+        resourceRegistration.registerReadOnlyAttribute(BUFFER_PER_SLICE, new BufferReadAttributeHandler(BUFFER_PER_SLICE));
+        resourceRegistration.registerReadOnlyAttribute(DIRECT_BUFFERS, new BufferReadAttributeHandler(DIRECT_BUFFERS));
+    }
+
+    private static class BufferReadAttributeHandler implements OperationStepHandler {
+
+        final SimpleAttributeDefinition definition;
+
+        public BufferReadAttributeHandler(SimpleAttributeDefinition definition) {
+            this.definition = definition;
+        }
 
-        private BufferPoolAdd() {
-            super(BufferPoolResourceDefinition.ATTRIBUTES);
+        private ModelNode returnConfigIfDefinedOrDefaultValueFor(String attributeName, ModelNode config, ModelNode defaultValue) {
+            System.out.println("RPE>>> config item is: " + config);
+            return (config != null && config.isDefined() && config.get(attributeName) != null
+                    && config.get(attributeName).isDefined() && !"".equals(config.get(attributeName))) ? config
+                    .get(attributeName) : defaultValue;
         }
 
         @Override
-        protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {
+        public void execute(OperationContext context, ModelNode operation) throws OperationFailedException {
+            // If values were defined in the subsystem...
             final PathAddress address = PathAddress.pathAddress(operation.get(OP_ADDR));
-            final String name = address.getLastElement().getValue();
-            final ModelNode bufferSizeModel = BUFFER_SIZE.resolveModelAttribute(context, model);
-            final ModelNode bufferPerSliceModel = BUFFER_PER_SLICE.resolveModelAttribute(context, model);
-            final ModelNode directModel = DIRECT_BUFFERS.resolveModelAttribute(context, model);
-
-            final int bufferSize = bufferSizeModel.isDefined() ? bufferSizeModel.asInt() : defaultBufferSize;
-            final int bufferPerSlice = bufferPerSliceModel.isDefined() ? bufferPerSliceModel.asInt() : defaultBuffersPerRegion;
-            final boolean direct = directModel.isDefined() ? directModel.asBoolean() : defaultDirectBuffers;
-
-            final BufferPoolService service = new BufferPoolService(bufferSize, bufferPerSlice, direct);
-            context.getCapabilityServiceTarget().addCapability(IO_POOL_RUNTIME_CAPABILITY, service)
-                    .setInitialMode(ServiceController.Mode.ACTIVE)
-                    .install();
+            final ModelNode config = Resource.Tools.readModel(context.readResourceFromRoot(address));
```

Comment:
code seems a bit convoluted.

Why don't you just call the `definition.resolveModelAttribute(context, model)`?
If it is defined, returns the value, else, returns one of the DEFAULT_XXX that you could pass to the class when you instantiate it with the corresponding attribute definition?

Reply:
...because I did not see this method! Thanks for the tip!